README
I consider Convert::ASN1 a replacement for my earlier Convert::BER
module. While ASN1.pm is not as flexable as BER.pm, because PDUs must
be described up front, it is also more powerful.  For example an LDAP
filter is a recursive structure, BER.pm cannot encode or decode this in
a single pass, ASN1.pm can.

Convert::ASN1 will parse ASN.1 descriptions and will encode from and
decode to perl data structures using a hierarchy of references.

Copyright (c) 2000-2002 Graham Barr. All rights reserved.
This package is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

The latest release is available from CPAN

  http://search.cpan.org/author/GBARR/Convert-ASN1

INSTALL

To install run these commands, substituting x.xx for the version number
that you have downloaded

  gunzip Convert-ASN1-x.xx.tar.gz
  tar xvf Convert-ASN1-x.xx.tar
  cd Convert-ASN1-x.xx
  perl Makefile.PL
  make
  make test
  make install

REPORTING BUGS

If you find any bugs with Convert::ASN1, please report then via

  bugs-Convert-ASN1@rt.cpan.org

$Id: README,v 1.4 2002/08/19 23:59:18 gbarr Exp $


ChangeLog
{{$NEXT}}

0.34 -- Mon Aug 07 19:46:22 ADT 2023

Bug Fixes

Thanks to chrisridd and fschlich for these bug fixes
  * Fixes: #47 2nd - Debug output using print instead of printf
  * Fixes #47 Debug output using print instead of printf
  * ASN1.pod: fix minor typo

0.33 -- Wed Sep 22 19:49:53 ADT 2021

  * Official CPAN release of 0.33

0.32 -- Tue Sep 21 18:45:20 ADT 2021

  * examples/x509decode: parameters are optional for AlgorithmIdentifier when using [ec]dsa algorithms
  * t/19issue14.t: issue with warnings
  * Fixes #44 Redundant argument in printf warning
  * examples/x509decode: fix prototype error
  * examples/ldap: fix asn1_dump not working

0.31 -- Wed Jun 02 22:28:29 ADT 2021

  * Official CPAN release of 0.30

0.30 -- Sat May 29 21:56:29 ADT 2021

  * Fixes #43 - Perl 5.35.1 makes scalar assignment to undef a compile time error

0.29 -- Mon May 24 18:27:46 ADT 2021

  * Release Version no changes from 0.28
  * Increment Version in Makefile.PL

0.28 -- Sun May 23 18:03:47 ADT 2021

  * Add github actions for repo testing and add new maintainer [Timothy Legge]
  * typo fix [Adam Leskis]
  * Fix test failures under Perl 5.26+ without '.' in @INC [Kent Fredric (KENTNL)]
  * Correct a typo - SYNOPSIS [Shlomi Fish]
  * unsafe decoding CVE-2013-7488 [Dana Jacobsen]
  * Typo fixes from dsteinbrunner [David Steinbrunner]
  * Add files via upload [Peter Sylvester]

0.27 -- Wed Jun 25 13:47:28 CDT 2014
  * Avoid recursion when encountering an invalid tag and length
  * Avoid negative repetition counts (5.22 compatibility) [Alex Vandiver]
  * Use pack/unpack instead of chr/org to avoid encoding issues

0.26 -- Sat Jun 09 13:30:16 CDT 2012
  * Avoid Modification of a read-only value attempted errors
  * dump Data::Dumper strings when rtest fails
  * Add tests from Crypt::X509 that found the last Modification of a read-only value attempted error

0.25 -- Fri Jun 08 19:31:59 CDT 2012
  * fix a spelling mistake [gregor herrmann]
  * Add use strict and fix errors
  * Fix decoding of indefinite sequences with ANY
  * POD additions for ANY
  * EXPLICIT ops must pass their cVAR to their child op
  * Swallow a , after a }

0.24 -- Mon Jun 04 17:11:08 CDT 2012
  * spelling fix [gregor herrmann]
  * Added tests for extension markers (fails currently) [Niels Laukens]
  * Bugfix for extension markerss in SETs and CHOICEs [Niels Laukens]
  * Skip attempting to encode opEXTENSIONS inside of CHOICE
  * Support for default explicit tagging [Peter Mogensen]
  * Added tests and doc for explicit tagging [Peter Mogensen]
  * Add new opEXPLICIT for sequences that are inserted for EXPLICIT tags

0.23 -- Thu May 03 16:32:46 CDT 2012
 * Added support for EXTENSION MARKERS (...) (Niels Laukens)
 * Switch to Dist::Zilla for building

Convert::ASN1 0.21 -- Mon Sep 15 14:34:45 CDT 2008

Bug Fixes
 * Fix decoding of OIDs with large value for 2nd element

Enhancements
 * Support for SEQUENCE with no elements in ASN.1
 * Added ability to define what value is put in perl data structory when decoding NULL elements

Convert::ASN1 0.21 -- Fri Feb  2 20:24:59 CST 2007

Bug Fixes
 * Fixed problem with indefinite encoding inside an indefinite encoding

Enhancements
 * Added support for type-local ANY DEFINED BY (patch from Leif Johansson)

Convert::ASN1 0.20 -- Tue Feb 21 18:47:43 CST 2006

Bug Fixes
 * Fixed issues with decoding when calling script contains use encoding 'utf8'
 * Allow zero-length timestamps to be extracted
 * Fix incorrect length encoding for bitstrings

Enhancements
 * Support encoding bigint's as BCD


Convert::ASN1 0.19 -- Mon Apr 18 19:40:32 CDT 2005

Bug Fixes
  * Fixed reading of packets with indefinite-length encodings

Enhancements
  * Add support for BCDString




2003-10-08 15:29  Graham Barr

	* lib/Convert/ASN1.pm:

	Release 0.18
	
2003-10-08 15:28  Graham Barr

	* MANIFEST, Makefile.PL, t/0-signature.t:

	Use Module::Build to build Makefile and add SIGNATURE
	
2003-10-08 13:28  Graham Barr

	* lib/Convert/: ASN1.pm, ASN1/_encode.pm:

	Fix bug in encoding BIT STRINGS where chr() was causing an upgrade to UTF8
	
2003-05-12 18:45  Graham Barr

	* lib/Convert/ASN1.pm:

	Release 0.17
	
2003-05-12 18:45  Graham Barr

	* MANIFEST:

	Add new tests into MANIFEST
	
2003-05-12 18:06  Graham Barr

	* t/funcs.pl:

	Skip rtest if Data::Dumper is not available
	
2003-05-07 16:13  Graham Barr

	* parser.y, lib/Convert/ASN1/parser.pm, t/04opt.t, t/funcs.pl:

	Support OPTIONAL on SET OF and SEQUENCE OF
	
2003-05-07 10:26  Graham Barr

	* lib/Convert/ASN1/_decode.pm, t/00prim.t:

	Fix OID decoding
	
2003-05-06 22:47  Graham Barr

	* t/14any.t:

	Remove duplicate my
	
2003-05-06 22:29  Graham Barr

	* parser.y, lib/Convert/ASN1.pm, lib/Convert/ASN1/_decode.pm,
	lib/Convert/ASN1/_encode.pm, lib/Convert/ASN1/parser.pm, t/14any.t:

	Add support for ANY DEFINED BY
	Patch from Simon Wilkinson
	
2003-05-06 15:17  Graham Barr

	* lib/Convert/ASN1/Debug.pm:

	Fix debug output for OIDs
	
2003-05-06 13:40  Graham Barr

	* parser.y, lib/Convert/ASN1.pm, lib/Convert/ASN1.pod,
	lib/Convert/ASN1/parser.pm, t/08set.t, t/12der.t, t/13utf8.t:

	CER and DER SET encoding
	
2003-05-06 12:07  Graham Barr

	* t/13utf8.t, lib/Convert/ASN1.pm, lib/Convert/ASN1/_decode.pm,
	lib/Convert/ASN1/_encode.pm:

	utf8 support for perl >= 5.8
	
2002-08-20 01:00  Graham Barr

	* lib/Convert/ASN1.pm:

	Release 0.16
	
2002-08-20 00:59  Graham Barr

	* README, lib/Convert/ASN1.pod:

	Update search.cpan.org link and add CPAN RT email address
	
2002-08-20 00:53  Graham Barr

	* t/10choice.t:

	Fix test count
	

t/funcs.pl

sub ntest ($$$) {
  my $ret = 1;
  if ($_[1] != $_[2]) {
    printf "#$_[0]: expecting $_[1]\n";
    printf "#$_[0]:       got $_[2]\n";
    printf "#line %d %s\n",(caller)[2,1];
    print "not ";
    $ret = 0;
  }
  print "ok $_[0]\n";
  $ret;
}

sub stest ($$$) {
  my $ret = 1;
  unless (defined $_[2] && $_[1] eq $_[2]) {
    printf "#$_[0]: expecting %s\n", $_[1] =~ /[^\.\d\w]/ ? "hex:".unpack("H*",$_[1]) : $_[1];
    printf "#$_[0]:       got %s\n", defined($_[2]) ? $_[2] =~  /[^\.\d\w]/ ? "hex:".unpack("H*",$_[2]) : $_[2] : 'undef';
    printf "#line %d %s\n",(caller)[2,1];
    print "not ";
    $ret = 0;
  }
  print "ok $_[0]\n";
  $ret;
}

sub btest ($$) {
  unless ($_[1]) {
    printf "#line %d %s\n",(caller)[2,1];
    print "not ";
  }
  print "ok $_[0]\n";
  $_[1]
}


sub rtest ($$$) {
  unless (eval { require Data::Dumper } ) {
    print "ok $_[0] # skip need Data::Dumper\n";
    return;
  }

  local $Data::Dumper::Sortkeys = 1;
  local $Data::Dumper::Useqq = 1;
  my $ok = Data::Dumper::Dumper($_[1]) eq Data::Dumper::Dumper($_[2]);

  unless ($ok) {
    printf "#line %d %s\n",(caller)[2,1];
    print "not ";
  }
  print "ok $_[0]\n";
  $ok;
}

1;



Makefile.PL
use strict;
use warnings;



use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Convert between perl data structures and ASN.1 encoded packets",
  "AUTHOR" => "Graham Barr <gbarr\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Convert-ASN1",
  "LICENSE" => "perl",
  "NAME" => "Convert::ASN1",
  "PREREQ_PM" => {},
  "TEST_REQUIRES" => {
    "Math::BigInt" => "1.997",
    "Test::More" => "0.90"
  },
  "VERSION" => "0.34",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Math::BigInt" => "1.997",
  "Test::More" => "0.90"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Convert/ASN1/IO.pm
# Copyright (c) 2000-2005 Graham Barr <gbarr@pobox.com>. All rights reserved.
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.

package Convert::ASN1;

use strict;
use Socket;

BEGIN {
  local $SIG{__DIE__};
  eval { require bytes } and 'bytes'->import
}

sub asn_recv { # $socket, $buffer, $flags

  my $peer;
  my $buf;
  my $n = 128;
  my $pos = 0;
  my $depth = 0;
  my $len = 0;
  my($tmp,$tb,$lb);

  MORE:
  for(
    $peer = recv($_[0],$buf,$n,MSG_PEEK);
    defined $peer;
    $peer = recv($_[0],$buf,$n<<=1,MSG_PEEK)
  ) {

    if ($depth) { # Are we searching of "\0\0"

      unless (2+$pos <= length $buf) {
	next MORE if $n == length $buf;
	last MORE;
      }

      if(substr($buf,$pos,2) eq "\0\0") {
	unless (--$depth) {
	  $len = $pos + 2;
	  last MORE;
	}
      }
    }

    # If we can decode a tag and length we can determine the length
    ($tb,$tmp) = asn_decode_tag(substr($buf,$pos));
    unless ($tb || $pos+$tb < length $buf) {
      next MORE if $n == length $buf;
      last MORE;
    }

    if (unpack("C",substr($buf,$pos+$tb,1)) == 0x80) {
      # indefinite length, grrr!
      $depth++;
      $pos += $tb + 1;
      redo MORE;
    }

    ($lb,$len) = asn_decode_length(substr($buf,$pos+$tb));

    if ($lb) {
      if ($depth) {
	$pos += $tb + $lb + $len;
	redo MORE;
      }
      else {
	$len += $tb + $lb + $pos;
	last MORE;
      }
    }
  }

  if (defined $peer) {
    if ($len > length $buf) {
      # Check we can read the whole element
      goto error
	unless defined($peer = recv($_[0],$buf,$len,MSG_PEEK));

      if ($len > length $buf) {
	# Cannot get whole element
	$_[1]='';
	return $peer;
      }
    }
    elsif ($len == 0) {
      $_[1] = '';
      return $peer;
    }

    if ($_[2] & MSG_PEEK) {
      $_[1] =  substr($buf,0,$len);
    }
    elsif (!defined($peer = recv($_[0],$_[1],$len,0))) {
      goto error;
    }

    return $peer;
  }

error:
    $_[1] = undef;
}

sub asn_read { # $fh, $buffer, $offset

  # We need to read one packet, and exactly only one packet.
  # So we have to read the first few bytes one at a time, until
  # we have enough to decode a tag and a length. We then know
  # how many more bytes to read

  if ($_[2]) {
    if ($_[2] > length $_[1]) {
      require Carp;
      Carp::carp("Offset beyond end of buffer");
      return;
    }
    substr($_[1],$_[2]) = '';
  }
  else {
    $_[1] = '';
  }

  my $pos = 0;
  my $need = 0;
  my $depth = 0;
  my $ch;
  my $n;
  my $e;
  

  while(1) {
    $need = ($pos + ($depth * 2)) || 2;

    while(($n = $need - length $_[1]) > 0) {
      $e = sysread($_[0],$_[1],$n,length $_[1]) or
	goto READ_ERR;
    }

    my $tch = unpack("C",substr($_[1],$pos++,1));
    # Tag may be multi-byte
    if(($tch & 0x1f) == 0x1f) {
      my $ch;
      do {
        $need++;
	while(($n = $need - length $_[1]) > 0) {
	  $e = sysread($_[0],$_[1],$n,length $_[1]) or
	      goto READ_ERR;
	}
	$ch = unpack("C",substr($_[1],$pos++,1));
      } while($ch & 0x80);
    }

    $need = $pos + 1;

    while(($n = $need - length $_[1]) > 0) {
      $e = sysread($_[0],$_[1],$n,length $_[1]) or
	  goto READ_ERR;
    }

    my $len = unpack("C",substr($_[1],$pos++,1));

    if($len & 0x80) {
      unless ($len &= 0x7f) {
	$depth++;
	next;
      }
      $need = $pos + $len;

      while(($n = $need - length $_[1]) > 0) {
	$e = sysread($_[0],$_[1],$n,length $_[1]) or
	    goto READ_ERR;
      }

      $pos += $len + unpack("N", "\0" x (4 - $len) . substr($_[1],$pos,$len));
    }
    elsif (!$len && !$tch) {
      die "Bad ASN PDU" unless $depth;
      unless (--$depth) {
	last;
      }
    }
    else {
      $pos += $len;
    }

    last unless $depth;
  }

  while(($n = $pos - length $_[1]) > 0) {
    $e = sysread($_[0],$_[1],$n,length $_[1]) or
      goto READ_ERR;
  }

  return length $_[1];

READ_ERR:
    $@ = defined($e) ? "Unexpected EOF" : "I/O Error $!"; # . CORE::unpack("H*",$_[1]);
    return undef;
}

sub asn_send { # $sock, $buffer, $flags, $to

  @_ == 4
    ? send($_[0],$_[1],$_[2],$_[3])
    : send($_[0],$_[1],$_[2]);
}

sub asn_write { # $sock, $buffer

  syswrite($_[0],$_[1], length $_[1]);
}

sub asn_get { # $fh

  my $fh = ref($_[0]) ? $_[0] : \($_[0]);
  my $href = \%{*$fh};

  $href->{'asn_buffer'} = '' unless exists $href->{'asn_buffer'};

  my $need = delete $href->{'asn_need'} || 0;
  while(1) {
    next if $need;
    my($tb,$tag) = asn_decode_tag($href->{'asn_buffer'}) or next;
    my($lb,$len) = asn_decode_length(substr($href->{'asn_buffer'},$tb,8)) or next;
    $need = $tb + $lb + $len;
  }
  continue {
    if ($need && $need <= length $href->{'asn_buffer'}) {
      my $ret = substr($href->{'asn_buffer'},0,$need);
      substr($href->{'asn_buffer'},0,$need) = '';
      return $ret;
    }

    my $get = $need > 1024 ? $need : 1024;

    sysread($_[0], $href->{'asn_buffer'}, $get, length $href->{'asn_buffer'})
      or return undef;
  }
}

sub asn_ready { # $fh

  my $fh = ref($_[0]) ? $_[0] : \($_[0]);
  my $href = \%{*$fh};

  return 0 unless exists $href->{'asn_buffer'};
  
  return $href->{'asn_need'} <= length $href->{'asn_buffer'}
    if exists $href->{'asn_need'};

  my($tb,$tag) = asn_decode_tag($href->{'asn_buffer'}) or return 0;
  my($lb,$len) = asn_decode_length(substr($href->{'asn_buffer'},$tb,8)) or return 0;

  $href->{'asn_need'} = $tb + $lb + $len;

  $href->{'asn_need'} <= length $href->{'asn_buffer'};
}

1;


lib/Convert/ASN1.pm
# Copyright (c) 2000-2002 Graham Barr <gbarr@pobox.com>. All rights reserved.
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.

package Convert::ASN1;

use 5.004;
use strict;
use vars qw($VERSION @ISA @EXPORT_OK %EXPORT_TAGS @opParts @opName $AUTOLOAD);
use Exporter;

use constant CHECK_UTF8 => $] > 5.007;

BEGIN {
  local $SIG{__DIE__};
  eval { require bytes and 'bytes'->import };

  if (CHECK_UTF8) {
    require Encode;
    require utf8;
  }

  @ISA = qw(Exporter);

  %EXPORT_TAGS = (
    io    => [qw(asn_recv asn_send asn_read asn_write asn_get asn_ready)],

    debug => [qw(asn_dump asn_hexdump)],

    const => [qw(ASN_BOOLEAN     ASN_INTEGER      ASN_BIT_STR      ASN_OCTET_STR
		 ASN_NULL        ASN_OBJECT_ID    ASN_REAL         ASN_ENUMERATED
		 ASN_SEQUENCE    ASN_SET          ASN_PRINT_STR    ASN_IA5_STR
		 ASN_UTC_TIME    ASN_GENERAL_TIME ASN_RELATIVE_OID
		 ASN_UNIVERSAL   ASN_APPLICATION  ASN_CONTEXT      ASN_PRIVATE
		 ASN_PRIMITIVE   ASN_CONSTRUCTOR  ASN_LONG_LEN     ASN_EXTENSION_ID ASN_BIT)],

    tag   => [qw(asn_tag asn_decode_tag2 asn_decode_tag asn_encode_tag asn_decode_length asn_encode_length)],
    
    hex   => [qw(hex2twos twos2hex)]
  );

  @EXPORT_OK = map { @$_ } values %EXPORT_TAGS;
  $EXPORT_TAGS{all} = \@EXPORT_OK;

  @opParts = qw(
    cTAG cTYPE cVAR cLOOP cOPT cEXT cCHILD cDEFINE
  );

  @opName = qw(
    opUNKNOWN opBOOLEAN opINTEGER opBITSTR opSTRING opNULL opOBJID opREAL
    opSEQUENCE opEXPLICIT opSET opUTIME opGTIME opUTF8 opANY opCHOICE opROID opBCD
    opEXTENSIONS
  );

  foreach my $l (\@opParts, \@opName) {
    my $i = 0;
    foreach my $name (@$l) {
      my $j = $i++;
      no strict 'refs';
      *{__PACKAGE__ . '::' . $name} = sub () { $j }
    }
  }
}

sub _internal_syms {
  my $pkg = caller;
  no strict 'refs';
  for my $sub (@opParts,@opName,'dump_op') {
    *{$pkg . '::' . $sub} = \&{__PACKAGE__ . '::' . $sub};
  }
}

sub ASN_BOOLEAN 	() { 0x01 }
sub ASN_INTEGER 	() { 0x02 }
sub ASN_BIT_STR 	() { 0x03 }
sub ASN_OCTET_STR 	() { 0x04 }
sub ASN_NULL 		() { 0x05 }
sub ASN_OBJECT_ID 	() { 0x06 }
sub ASN_REAL 		() { 0x09 }
sub ASN_ENUMERATED	() { 0x0A }
sub ASN_RELATIVE_OID	() { 0x0D }
sub ASN_SEQUENCE 	() { 0x10 }
sub ASN_SET 		() { 0x11 }
sub ASN_PRINT_STR	() { 0x13 }
sub ASN_IA5_STR		() { 0x16 }
sub ASN_UTC_TIME	() { 0x17 }
sub ASN_GENERAL_TIME	() { 0x18 }

sub ASN_UNIVERSAL 	() { 0x00 }
sub ASN_APPLICATION 	() { 0x40 }
sub ASN_CONTEXT 	() { 0x80 }
sub ASN_PRIVATE		() { 0xC0 }

sub ASN_PRIMITIVE	() { 0x00 }
sub ASN_CONSTRUCTOR	() { 0x20 }

sub ASN_LONG_LEN	() { 0x80 }
sub ASN_EXTENSION_ID	() { 0x1F }
sub ASN_BIT 		() { 0x80 }


sub new {
  my $pkg = shift;
  my $self = bless {}, $pkg;

  $self->configure(@_);
  $self;
}


sub configure {
  my $self = shift;
  my %opt = @_;

  $self->{options}{encoding} = uc($opt{encoding} || 'BER');

  unless ($self->{options}{encoding} =~ /^[BD]ER$/) {
    require Carp;
    Carp::croak("Unsupported encoding format '$opt{encoding}'");
  }

  # IMPLICIT as defalt for backwards compatibility, even though it's wrong.
  $self->{options}{tagdefault} = uc($opt{tagdefault} || 'IMPLICIT');

  unless ($self->{options}{tagdefault} =~ /^(?:EXPLICIT|IMPLICIT)$/) {
    require Carp;
    Carp::croak("Default tagging must be EXPLICIT/IMPLICIT. Not $opt{tagdefault}");
  }


  for my $type (qw(encode decode)) {
    if (exists $opt{$type}) {
      while(my($what,$value) = each %{$opt{$type}}) {
	$self->{options}{"${type}_${what}"} = $value;
      }
    }
  }
}



sub find {
  my $self = shift;
  my $what = shift;
  return unless exists $self->{tree}{$what};
  my %new = %$self;
  $new{script} = $new{tree}->{$what};
  bless \%new, ref($self);
}


sub prepare {
  my $self = shift;
  my $asn  = shift;

  $self = $self->new unless ref($self);
  my $tree;
  if( ref($asn) eq 'GLOB' ){
    local $/ = undef;
    my $txt = <$asn>;
    $tree = Convert::ASN1::parser::parse($txt,$self->{options}{tagdefault});
  } else {
    $tree = Convert::ASN1::parser::parse($asn,$self->{options}{tagdefault});
  }

  unless ($tree) {
    $self->{error} = $@;
    return;
    ### If $self has been set to a new object, not returning
    ### this object here will destroy the object, so the caller
    ### won't be able to get at the error.
  }

  $self->{tree} = _pack_struct($tree);
  $self->{script} = (values %$tree)[0];
  $self;
}

sub prepare_file {
  my $self = shift;
  my $asnp = shift;

  local *ASN;
  open( ASN, $asnp )
      or do{ $self->{error} = $@; return; };
  my $ret = $self->prepare( \*ASN );
  close( ASN );
  $ret;
}

sub registeroid {
  my $self = shift;
  my $oid  = shift;
  my $handler = shift;

  $self->{options}{oidtable}{$oid}=$handler;
  $self->{oidtable}{$oid}=$handler;
}

sub registertype {
   my $self = shift;
   my $def = shift;
   my $type = shift;
   my $handler = shift;

   $self->{options}{handlers}{$def}{$type}=$handler;
}

# In XS the will convert the tree between perl and C structs

sub _pack_struct { $_[0] }
sub _unpack_struct { $_[0] }

##
## Encoding
##

sub encode {
  my $self  = shift;
  my $stash = @_ == 1 ? shift : { @_ };
  my $buf = '';
  local $SIG{__DIE__};
  eval { _encode($self->{options}, $self->{script}, $stash, [], $buf) }
    or do { $self->{error} = $@; undef }
}



# Encode tag value for encoding.
# We assume that the tag has been correctly generated with asn_tag()

sub asn_encode_tag {
  $_[0] >> 8
    ? $_[0] & 0x8000
      ? $_[0] & 0x800000
	? pack("V",$_[0])
	: substr(pack("V",$_[0]),0,3)
      : pack("v", $_[0])
    : pack("C",$_[0]);
}


# Encode a length. If < 0x80 then encode as a byte. Otherwise encode
# 0x80 | num_bytes followed by the bytes for the number. top end
# bytes of all zeros are not encoded

sub asn_encode_length {

  if($_[0] >> 7) {
    my $lenlen = &num_length;

    return pack("Ca*", $lenlen | 0x80,  substr(pack("N",$_[0]), -$lenlen));
  }

  return pack("C", $_[0]);
}


##
## Decoding
##

sub decode {
  my $self  = shift;
  my $ret;

  local $SIG{__DIE__};
  eval {
    my (%stash, $result);
    my $script = $self->{script};
    my $stash = \$result;

    while ($script) {
      my $child = $script->[0] or last;
      if (@$script > 1 or defined $child->[cVAR]) {
        $result = $stash = \%stash;
        last;
      }
      last if $child->[cTYPE] == opCHOICE or $child->[cLOOP];
      $script = $child->[cCHILD];
    }

    _decode(
	$self->{options},
	$self->{script},
	$stash,
	0,
	length $_[0], 
	undef,
	{},
	$_[0]);

    $ret = $result;
    1;
  } or $self->{'error'} = $@ || 'Unknown error';

  $ret;
}


sub asn_decode_length {
  return unless length $_[0];

  my $len = unpack("C",$_[0]);

  if($len & 0x80) {
    $len &= 0x7f or return (1,-1);

    return if $len >= length $_[0];

    return (1+$len, unpack("N", "\0" x (4 - $len) . substr($_[0],1,$len)));
  }
  return (1, $len);
}


sub asn_decode_tag {
  return unless length $_[0];

  my $tag = unpack("C", $_[0]);
  my $n = 1;

  if(($tag & 0x1f) == 0x1f) {
    my $b;
    do {
      return if $n >= length $_[0];
      $b = unpack("C",substr($_[0],$n,1));
      $tag |= $b << (8 * $n++);
    } while($b & 0x80);
  }
  ($n, $tag);
}


sub asn_decode_tag2 {
  return unless length $_[0];

  my $tag = unpack("C",$_[0]);
  my $num = $tag & 0x1f;
  my $len = 1;

  if($num == 0x1f) {
    $num = 0;
    my $b;
    do {
      return if $len >= length $_[0];
      $b = unpack("C",substr($_[0],$len++,1));
      $num = ($num << 7) + ($b & 0x7f);
    } while($b & 0x80);
  }
  ($len, $tag, $num);
}


##
## Utilities
##

# How many bytes are needed to encode a number 

sub num_length {
  $_[0] >> 8
    ? $_[0] >> 16
      ? $_[0] >> 24
	? 4
	: 3
      : 2
    : 1
}

# Convert from a bigint to an octet string

sub i2osp {
    my($num, $biclass) = @_;
    if (ref($num) eq 'Math::BigInt') {
      return hex2os( $num->as_hex($num) );
    }
    eval "use $biclass";
    $num = $biclass->new($num);
    my $neg = $num < 0
      and $num = abs($num+1);
    my $base = $biclass->new(256);
    my $result = '';
    while($num != 0) {
        my $r = $num % $base;
        $num = ($num-$r) / $base;
        $result .= pack("C",$r);
    }
    $result ^= pack("C",255) x length($result) if $neg;
    return scalar reverse $result;
}

# Convert from an octet string to a bigint

sub os2ip {
    my($os, $biclass) = @_;
    eval "require $biclass";
    if ($biclass eq 'Math::BigInt') {
      return $biclass->new( os2hex($os) );
    }
    my $base = $biclass->new(256);
    my $result = $biclass->new(0);
    my $neg = unpack("C",$os) >= 0x80
      and $os ^= pack("C",255) x length($os);
    for (unpack("C*",$os)) {
      $result = ($result * $base) + $_;
    }
    return $neg ? ($result + 1) * -1 : $result;
}

# Convert from a twos complement octet string to a signed hex string

sub os2hex {
    my $os = $_[0];
    my $len = length($os);
    # deterime if number is negative
    my $neg = unpack("C",$os) >= 0x80;
    if ($neg) {
      $os ^= "\xff" x $len;
      for(my$l=$len-1;$l>=0;$l--) {
        vec($os,$l,8) = vec($os,$l,8) + 1;
        last if vec($os,$l,8);
      }
    }
    # trim unnecessary leading zeros
    $os =~ s/^\x00+(?!$)//;
    my $result = $neg ? "-0x" : "0x";
    $result .= unpack("H*",$os);
    return $result;
}

# Convert from a signed hex string to a twos complement octet string

sub hex2os {
    my $num = $_[0];
    my $neg = substr($num,0,1) eq '-';
    my $hx = substr($num,$neg ? 3 : 2);
    my $pad = (length($hx) % 2) ? '0' : '';
    my $os = pack("H*",$pad . $hx);
    # trim to minimum size
    $os =~ s/^\x00+(?!$)//;
    return $os if $os eq "\x00";
    my $len = length $os;
    # 2s complement
    if ($neg) {
      $os ^= "\xff" x $len;
      for(my$l=$len-1;$l>=0;$l--) {
        vec($os,$l,8) = vec($os,$l,8) + 1;
        last if vec($os,$l,8);
      }
    }
    # verify most significant bit is properly signed
    my $msb = (vec($os,0,8) & 0x80) ? 0 : 255;
    $os = pack("C",$msb) . $os if $msb xor !$neg;
    return $os;
}

# Convert from a signed hex string to a twos complement hex string

sub hex2twos {
    my $os = hex2os($_[0]);
    return '0x'.unpack('H*',$os);
}

# Convert from a twos complement hex string to a signed hex string

sub twos2hex {
    my $s = $_[0] =~ m/^0x/i ? 2 : 0;
    my $os = pack('H*',substr($_[0],$s));
    return os2hex($os);
}

# Given a class and a tag, calculate an integer which when encoded
# will become the tag. This means that the class bits are always
# in the bottom byte, so are the tag bits if tag < 30. Otherwise
# the tag is in the upper 3 bytes. The upper bytes are encoded
# with bit8 representing that there is another byte. This
# means the max tag we can do is 0x1fffff

sub asn_tag {
  my($class,$value) = @_;

  die sprintf "Bad tag class 0x%x",$class
    if $class & ~0xe0;

  unless ($value & ~0x1f or $value == 0x1f) {
    return (($class & 0xe0) | $value);
  }

  die sprintf "Tag value 0x%08x too big\n",$value
    if $value & 0xffe00000;

  $class = ($class | 0x1f) & 0xff;

  my @t = ($value & 0x7f);
  unshift @t, (0x80 | ($value & 0x7f)) while $value >>= 7;
  unpack("V",pack("C4",$class,@t,0,0));
}


BEGIN {
  # When we have XS &_encode will be defined by the XS code
  # so will all the subs in these required packages 
  unless (defined &_encode) {
    require Convert::ASN1::_decode;
    require Convert::ASN1::_encode;
    require Convert::ASN1::IO;
  }

  require Convert::ASN1::parser;
}

sub AUTOLOAD {
  require Convert::ASN1::Debug if $AUTOLOAD =~ /dump/;
  goto &{$AUTOLOAD} if defined &{$AUTOLOAD};
  require Carp;
  my $pkg = ref($_[0]) || ($_[0] =~ /^[\w\d]+(?:::[\w\d]+)*$/)[0];
  if ($pkg and UNIVERSAL::isa($pkg, __PACKAGE__)) { # guess it was a method call
    $AUTOLOAD =~ s/.*:://;
    Carp::croak(sprintf q{Can't locate object method "%s" via package "%s"},$AUTOLOAD,$pkg);
  }
  else {
    Carp::croak(sprintf q{Undefined subroutine &%s called}, $AUTOLOAD);
  }
}

sub DESTROY {}

sub error { $_[0]->{error} }
1;
